emulator/src/main/java/com/scalar/client/tool/emulator/ContractManagerEmulator.java

```diff
@@ -0,0 +1,159 @@
+import com.scalar.ledger.ledger.Ledger;
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javassist.ByteArrayClassPath;
+import javassist.ClassPool;
+import javassist.CtClass;
+import javassist.CtField;
+import javassist.CtMethod;
+import javassist.CtNewMethod;
+import javassist.Modifier;
+import javax.json.Json;
+import javax.json.JsonObject;
+
+public class ContractManagerEmulator {
+
+  private final ContractRegistry registry;
+  private final Map<String, Contract> cache;
+  public ContractManagerEmulator that;
+
+  public ContractManagerEmulator(ContractRegistry registry) {
+    this.registry = registry;
+    this.cache = new HashMap<String, Contract>();
+    removeContractClassInvokeMethodModifier();
+    that = this;
+  }
+
+  public void register(com.scalar.ledger.contract.ContractEntry entry) {
+    registry.bind(entry);
+  }
+
+  public void register(String id, String name, File file, JsonObject properties) {
+    try {
+      byte[] contract = Files.readAllBytes(file.toPath());
+      register(toContractEntry(id, name, contract, properties));
+    } catch (IOException e) {
+      throw new RegistryIOException("could not register contract " + id);
+    }
+  }
+
+  /**
+   * invoke(id, ledger, argument) in Contract class has modifier `final`. To override it for
+   * registered contracts, we need to reset the modifier.
+   */
+  private void removeContractClassInvokeMethodModifier() {
+    ClassPool pool = ClassPool.getDefault();
+    try {
+      CtClass clazz = pool.get("com.scalar.ledger.contract.Contract");
+      for (CtMethod method : clazz.getMethods()) {
+        if (method
+            .getLongName()
+            .equals(
+                "com.scalar.ledger.contract.Contract.invoke(java.lang.String,com.scalar.ledger.ledger.Ledger,javax.json.JsonObject)")) {
+          method.setModifiers(Modifier.PUBLIC);
+          break;
+        }
+      }
+      clazz.toClass();
+      clazz.defrost();
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  private ContractEntry toContractEntry(
+      String id, String name, byte[] contract, JsonObject properties) {
+    return new ContractEntry(
+        id,
+        name,
+        "holder_id",
+        1,
+        contract,
+        properties,
+        System.currentTimeMillis(),
+        "signature".getBytes());
+  }
+
+  public ContractEntry get(ContractEntry.Key key) {
+    return registry.lookup(key.getId());
+  }
+
+  /** Thes method is used to delegate contract's invoke. */
+  protected JsonObject delegatedInvoke(String id, Ledger ledger, JsonObject argument) {
+    try {
+      Field field = this.getClass().getField("that");
+      Object o = field.get(this);
+      Method m = o.getClass().getDeclaredMethod("getInstance", String.class);
```

Comment:
I'm not fully sure, but it can't simply call `getInstance` of this class because it is invoked from the hacked object ?

Reply:
Yes.
Javassist just overrides the bytecode of `invoke` of `.class` file by using the bytecode of `emulateInvoke` of `ContractManagerEmulator.class`.
Therefore, when the instance is created and the overridden bytecode is executed,
the `this` pointer is actually addressing the instance of the contract but not the instance of `ContractManagerEmulator`, so JVM can not locate the bytecode of `getInstance` (this.getInstance) function.

Reply:
So I have to hackly set a `that` pointer addressing `ContractManagerEmulator` instance.